Mage/src/main/java/mage/abilities/AbilityImpl.java

```diff
@@ -1074,6 +1074,11 @@ public boolean canChooseTarget(Game game, UUID playerId) {
     }
 
     protected static boolean canChooseTargetAbility(Ability ability, Modes modes, Game game, UUID controllerId) {
+        if (ability.getTargetAdjuster() != null){
+            // We can't easily determine how the adjuster will adjust the target
+            // So just always treat it as legal to cast
+            return true;
```

Comment:
Looks bad. Target adjusting must be called on playable/activate check and must setup correct targets already.

Looks like it's old discord's discuss and I recommended to fix it by canActivate -- did you try it?

![shot_240829_124507](https://github.com/user-attachments/assets/d6168ae0-66ae-4bc3-9431-1c5b7b1d254a)

![shot_240829_124517](https://github.com/user-attachments/assets/54334549-d5c7-477c-8c32-dcdb4af9dd91)

![shot_240829_124529](https://github.com/user-attachments/assets/6abbb079-e233-4c50-a92b-d5de9fde5be2)

Reply:
Sorry, should've explained myself further.

With Long River's Pull, the decision to give the gift has not yet been made in the canPlay check. Given that, even if we do call the adjuster, it wouldn't adjust the target correctly since it would default to "no gift". Sometimes the adjuster makes casting more restrictive, sometimes it makes casting less restrictive, so there's no default I can always use.

The only way I could think of to solve the issue would be to have any target adjuster that can make casting less restrictive to have a new function that return all of the possible Targets it could make, and then check if any of them succeed. That would be a lot more code, and would require adding that method to some pre-existing target adjusters that aren't generic.

Reply:
"max possible target" is not necessarily well-defined as a single Target. It might be for all existing cards (haven't checked), but for example:

"Destroy target creature. If ~ was kicked, instead destroy two target noncreature permanents,"

AFAIK there's no possible way to correctly handle this with one Target. I suppose you could use just "Target permanent"? Require the user of the target adjuster to define what the playable state target is? That's probably good enough for most situations, I guess.

Reply:
Yes, in many use cases getNetMana return `{any}` instead complex calculations. It's ok for UX. Show wrongly playable card is much better (AI and user has a choice) than show wrongly non playable card (AI and user can't play it at all).

Reply:
This PR is equivalent to always returning {any} for all target adjusters. I'll try the more complex solution later.

Reply:
It’s more equal to old ALLOW_USERS_TO_PUT_NON_PLAYABLE_SPELLS_ON_STACK_WORKAROUND )))

Reply:
This only skips the "is there a legal target" check, all other checks remain in place. Many previous target adjusters effectively did this, with no target originally and adding either form of the target in the adjuster.